### ISSUE TITLE
Using Internal Data with Gradient Tape

### ISSUE THREAD

**Issue by seanjkanderson** [GPflow/GPflow#1791]
I looked through various references including the [intro to GPFlow2](https://gpflow.readthedocs.io/en/master/notebooks/intro_to_gpflow2.html), but am not quite understanding how/if I can use Internal Data with Gradient Tape. This might be too much of a casadi question. In general though, is there a way to use SGPR with external data?

In particular, I'm trying to use GPFlow with CasADI for constrained numerical optimization. This functionality exists for CasADI although documentation is near non-existent for use with GPFlow2. In general, after defining a callback function for CasADI where the method eval(arg) takes in the optimization variables at each step, I can use Gradient Tape to get my objective value and gradient. 

This assumes that my objective is a callable that takes in ``xf_tensor`` and returns ``value_cost`` in `DemoCost` as in the example below:

```
import casadi as ca
import tensorflow as tf
from casadi import Sparsity
import gpflow
import numpy as np


def demo_cost(x):
    # assume that x=[x0,x1,u]
    x1 = x[0, 0::3]
    x2 = x[0, 1::3]
    xp = tf.concat([x1, x2], axis=0)
    xp = tf.reshape(xp, (-1,1))
    u = tf.reshape(x[0, 2::3], (-1, 1))
    u_ref = -1*tf.ones((nd-1, 1), dtype=tf.dtypes.float64)
    du = u - u_ref
    return tf.transpose(xp) @ xp * 0 + tf.transpose(du)  @ du


class DemoCost(TensorFlowEvaluator):
    def __init__(self, model, opts:dict=None):
        if opts is None:
            opts = {}

        initializer = tf.random_normal_initializer(mean=1., stddev=2.)
        X = tf.Variable(initializer(shape=[1, 62], dtype=tf.float64), trainable=False)

        @tf.function
        def f_k(input_dat, get_grad_val=None):
            xf_tensor = input_dat
            if get_grad_val is not None:
                with tf.GradientTape(persistent=True) as tape:
                    tape.watch(xf_tensor)
                    value_cost = demo_cost(xf_tensor)
                grad_cost = tape.gradient(value_cost, xf_tensor, output_gradients=get_grad_val)
            else:
                value_cost = demo_cost(xf_tensor)
                grad_cost = None

            return value_cost, grad_cost

        TensorFlowEvaluator.__init__(self, [X], [f_k], model, opts=opts)
        self.counter = 0
        self.time = 0

    def eval(self, arg):
        self.counter += 1
        t0 = pytime.time()
        ret = TensorFlowEvaluator.eval(self, arg)
        self.time += pytime.time() - t0
        return [ret]


class TensorFlowEvaluator(ca.Callback):
    def __init__(self, t_in, t_out, model, set_init=False, opts={}):

        self.set_init = set_init
        self.opts = opts
        ca.Callback.__init__(self)
        assert isinstance(t_in, list)
        self.t_in = t_in
        assert isinstance(t_out, list)
        self.t_out = t_out
        self.output_shapes = []
        self.construct("TensorFlowEvaluator", {})
        self.refs = []
        self.model = model

    def get_n_in(self):
        return len(self.t_in)

    def get_n_out(self):
        return len(self.t_out)

    def get_sparsity_in(self, i):
        tesnor_shape = self.t_in[i].shape
        return Sparsity.dense(tesnor_shape[0], tesnor_shape[1])

    def get_sparsity_out(self, i):
        if (i == 0 and self.set_init is False):
            tensor_shape = [self.opts["output_dim"][0], self.opts["output_dim"][1]]
        elif (i == 0 and self.set_init is True):
            tensor_shape = [self.opts["grad_dim"][0], self.opts["grad_dim"][1]]
        else:
            tensor_shape = [self.opts["output_dim"][0], self.opts["output_dim"][1]]
        return Sparsity.dense(tensor_shape[0], tensor_shape[1])

    def eval(self, arg):
        updated_t = []
        for i, v in enumerate(self.t_in):
            updated_t.append(tf.Variable(arg[i].toarray()))
        if len(updated_t) == 1:
            out_, grad_estimate = self.t_out[0](tf.convert_to_tensor(updated_t[0]))
        else:
            out_, grad_estimate = self.t_out[0](tf.convert_to_tensor(updated_t[0]), tf.convert_to_tensor(updated_t[1]))

        if len(updated_t) == 1:
            selected_set = out_.numpy()
        else:
            selected_set = grad_estimate.numpy()
        return [selected_set]

    # Vanilla tensorflow offers just the reverse mode AD
    def has_reverse(self, nadj):
        return nadj == 1

    def get_reverse(self, nadj, name, inames, onames, opts):
        initializer = tf.random_normal_initializer(mean=1., stddev=2.)
        adj_seed = [tf.Variable(initializer(shape=self.sparsity_out(i).shape, dtype=tf.float64)) for i in
                    range(self.n_out())]

        callback = TensorFlowEvaluator(self.t_in + adj_seed, [self.t_out[0]], self.model, set_init=True, opts=self.opts)
        self.refs.append(callback)

        nominal_in = self.mx_in()
        nominal_out = self.mx_out()
        adj_seed = self.mx_out()
        casadi_bal = callback.call(nominal_in + adj_seed)
        return ca.Function(name, nominal_in + nominal_out + adj_seed, casadi_bal, inames, onames)
```

In my case, I would like to optimize the negative lower bound of the SGRP from Titsias et al. (2008). In the SGPR class, this seems to correspond to elbo() or maximum_log_likelihood_objective() which use the internal data paradigm. Since neither take an argument, I'm unsure of how to implement this in such a way as to mimic the requirement that TensorFlowEvaluator.eval() takes in the optimization variables at each step.

Thanks!

**Comment by jesnie** [comment]
What are you trying to achieve?

I assume you want to replace `demo_cost` with something with GPflow?

What is it you want to optimize? Normally we optimize the model hyperparameters, but since you're talking about internal data, is it the model data you want to optimize? The SGPR also has inducing points, which may make it harder if you're trying to optimize your data? Maybe it's the inducing points you want to optimize? Or maybe you should use a regular GPR instead of an SGPR?

**Comment by jesnie** [comment]
It occurs to me that, depending on what you're trying to accomplish, our SciPy optimizer may work as an example for you: https://github.com/GPflow/GPflow/blob/develop/gpflow/optimizers/scipy.py
It uses a `GradientTape` to get the gradients of the loss function with respect to all trainable parameters, typically the model hyper parameters.

**Comment by seanjkanderson** [comment]
Thanks @jesnie.  I ended up using some of the methods from SciPy to construct an objective function that has the format I need. I'm trying to optimize the location of the inducing variables in a constrained (trajectory) optimization setting. More specifically, I'm using SGPR.elbo() as the objective, and then hoping to solve a trajectory optimization using CasADI and a nonlinear solver.

I did have to create a child class that overloads a couple of the SciPy methods to make it general enough. I'm guessing not enough people have a need for a general Optimizer class to abstract SciPy to something more general, but if I do come up with some nice abstractions, I'll share those later.
